core/trino-main/src/main/java/io/trino/operator/Driver.java

```diff
@@ -526,6 +532,9 @@ private void destroyIfNecessary()
                     "Error destroying driver for task %s",
                     driverContext.getTaskId());
         }
+        finally {
+            destroyedFuture.set(null);
```

Comment:
So... I guess it's possible to move this outside of the locked region, but that will mean doing something like this in 3 other call sites:
```
finally {
  exclusiveLock.unlock();
  if (state.get() == DESTROYED) {
    destroyedFuture.set(null);
  }
}
```

Which seems more awkward to me in that it's spatially separating the action of destruction from setting the future. I Since right now that future is only tied to detecting termination at the task level, I think this is safe since the potential for deadlock is on the driver's exclusive lock and not on the any task-level lock- but it is riskier if the usage expands later on... maybe instead we don't expose `destroyedFuture` but instead allow passing a `Runnable` to the driver constructor directly? It's still the same thing as the cleanup action would run while holding the lock, but it seems less risky longer term since there's no opportunity to pack more listeners onto the driver later. Let me know what you think and I'll either keep what's there, move the future setting out of the lock region, or pass a cleanup runnable to the driver constructor instead.

Reply:
Update on the above: you actually can’t move it out of the exclusive lock region without potentially missing the final DriverContext finished() call, since failing to acquire the lock could still see the state flip to DESTROYED before the lock holder completes destruction.

This could be addressed by adding a new DESTROYING state, but that’s extra complexity too.

Reply:
Yeah, i can totally see how that would make the code less reasonable. I also agree that chances of introducing a deadlock here are rather low. However to me it still feels worth a decrease in readability, but I don't feel particularly strongly about it.

Reply:
I have the same concerns.  This code has a long history with deadlocks.  Maybe add a utility method like `checkDestroyedState`.
BTW a runnable won't work either as executing it while holding the lock is still an open out call.

Reply:
After looking over some of the options, it seemed to me the best compromise between avoiding potential deadlocks and not having multiple checks in different codepaths to decide when to fire the `destroyedFuture` is to add `Driver.State.DESTROYING` to separate "destruction in progress by some thread" and "destruction completed", and have `DriverLock#unlock()` complete the future immediately after it's released- so I've published that version in the latest revision.